pkg/algorithms/cmr/cmr.py

```diff
@@ -0,0 +1,123 @@
+
+        Args:
+            tolerancy (int): tolerancy to compare on 
+            metric_columns (List[str]): string list of metric column names
+            dataframe_list (pd.DataFrame): data frame of all data to compare on
+
+        Returns:
+            pd.Dataframe, dict[metric_name, ChangePoint]: Returned data frame and change points
+        """
+        change_points_by_metric={ k:[] for k in metric_columns }
+        max_date_time = pd.Timestamp.max.to_pydatetime()
+        max_time = max_date_time.timestamp()
+        difference = ["difference", max_time]
+        pass_fail_list = ["Pass/Fail", max_time]
+        for column in metric_columns:
+            pct_change_result = dataframe_list[column].pct_change()
+            single_pct_diff = round(pct_change_result.iloc[[-1]].values[0] * 100)
+            pass_fail = "Pass"
+            if single_pct_diff > tolerancy:
+                pass_fail = "Fail"
+
+            change_point = ChangePoint(metric=column,
+                                            index=1,
+                                            time=difference[1],
+                                            stats=ComparativeStats(
+                                                mean_1=dataframe_list[column][0],
+                                                mean_2=dataframe_list[column][1],
+                                                std_1=0,
+                                                std_2=0,
+                                                pvalue=1
+                                            ))
+            change_points_by_metric[column].append(change_point)
+            difference.append(single_pct_diff)
+            pass_fail_list.append(pass_fail)
+        # difference.append("none")
+        # pass_fail_list.append("none")
+        # dataframe_list.loc[len(dataframe_list.index)] = difference
+        #dataframe_list.loc[len(dataframe_list.index)] = pass_fail_list
+        # logger_instance.info("final data frame " + str(dataframe_list))
+
+        # based on change point generate pass/fail
+        return dataframe_list, change_points_by_metric
+
+    def combine_data_frames(self, dataFrame: pd.DataFrame):
+        """
+        If more than 1 previous run, mean data together into 1 single row
+        Combine with current run into 1 data frame (current run being -1 index)
+
+        Args:
+            dataFrame (pd.DataFrame): data to combine into 2 rows
+
+        Returns:
+            pd.Dataframe: data frame of most recent run and averaged previous runs
+        """
+        i = 0
+
+        last_row = dataFrame.tail(1)
+        dF = dataFrame[:-1]
+        data2 = {}
+
+        metric_columns = list(dataFrame.columns)
+        for column in metric_columns:
+
+            if isinstance(dF.loc[0, column], (numpy.float64, numpy.int64)):
+                mean = dF[column].mean()
+            else:
+                column_list = dF[column].tolist()
+                mean = ','.join(column_list)
```

Comment:
Do we really need this mean for non-numeric columns?

Reply:
It doesn't take a mean, it combines into a comma separated list that we're able to get the list of all the uuids that were meaned together and their build urls

Reply:
I was referring towards to variable `mean = ','.join(column_list)` and the purpose of doing it. Where are we using those combined values?

Reply:
If you look at the output [here](https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/scale-ci/job/paige-e2e-multibranch/job/orion/39/artifact/orion/output_aws-small-scale-node-density-cni.txt), its to see all the uuids that were used to generate the mean. If not we wouldnt have a way of knowing which data points were used when calculating the mean.

I used `mean` to be able to add either the non-numeric and numeric columns without having it in the if statement

Reply:
I totally understand the thought process behind the logic of having mean, but I was curious if we can move data2[column] inside the if else statement and rename mean in the else so that there is no confusion with the naming.